File: perceptron.py
import numpy as np

x = [1, 1, -1, -1]
y = [1, -1, -1, 1]

R1 = [1, -1, -1, -1]
R3 = [1, 1, 1, -1]
R2 = [-1, -1, 1, -1]
print("Choose options ")
print("1.AND")
print("2.OR")
print("3.ANDNOT")
option = input()
b = input("Enter the value of b")
w1 = input("Enter the value of w1")
w2 = input("Enter the value of w2")
theta = input("Enter the value of Theta ")
alpha = input("Enter the value of alpha")


def solve(x, y, R1, b, w1, w2, theta, alpha):
    epoch = 1
    f = False
    while(f == False and epoch < 5):
        f = True
        print("Epoch", end=" ")
        print(epoch)
        for i in range(0, 4):
            yin = b
            print(x[i])
            print(y[i])
            yin = yin+x[i]*w1+y[i]*w2
            print(yin)
            print("Yin for ", end=" ")
            print(i, end=" ")
            print("th pair", end=" ")
            print(yin)
            # apply activation
            Y = 0
            if(yin > theta):
                Y = 1
            else:
                Y = -1

            if(Y != R1[i]):
                w1 = w1+alpha*x[i]*R1[i]
                w2 = w2+alpha*y[i]*R1[i]
                b = b+alpha*R1[i]
                f = False
            print("After", end=" ")
            print(i, end=" ")
            print("th pair w1 ", end=" ")
            print(w1, end=" ")
            print("w2 ", end=" ")
            print(w2, end=" ")
            print("b", end=" ")
            print(b)
        print("After", end=" "),
        print(epoch, end=" "),
        print("th w1", end=" ")
        print(w1, end=" ")
        print("w2", end=" ")
        print(w2, end=" ")
        print("b", end=" ")
        print(b)
        epoch = epoch+1
    return


if(option == "AND"):
    solve(x, y, R1, b, w1, w2, theta, alpha)
elif(option == 'OR'):
    solve(x, y, R3, b, w1, w2, theta, alpha)
else:
    solve(x, y, R2, b, w1, w2, theta, alpha)

File: perceptron.cpp
#include <bits/stdc++.h>
using namespace std;
int x[4] = {1, -1, 1, -1};
int y[4] = {1, 1, -1, -1};
int R1[4] = {1, -1, -1, -1};
int R2[4] = {-1, -1, 1, -1};
int R3[4] = {1, 1, 1, -1};
int R4[4] = {-1, 1, 1, 1};
void solve(int *x, int *y, int *R1, int theta, int alpha, int b, int w1, int w2)
{
    int epoch = 1;
    bool f = false;
    while (!f)
    {
        f = true;
        cout << "epoch " << epoch << endl;
        for (int i = 0; i < 4; i++)
        {
            int yin = b;
            yin += w1 * x[i] + w2 * y[i];
            cout << "Yin for " << i << "th pair " << yin << endl;
            //apply activation function
            int Y = 0;
            if (yin > theta)
                Y = 1;
            else if (yin < theta)
                Y = -1;
            cout << "Y for " << i << "th pair " << Y << endl;
            if (Y != R1[i])
            {
                w1 = w1 + alpha * R1[i] * x[i];
                w2 = w2 + alpha * R1[i] * y[i];
                b += alpha * R1[i];
                f = false;
            }
            cout << "after " << i << "th pair w1 " << w1 << " w2 " << w2 << " b " << b << endl;
        }
        cout << "after " << epoch << "th w1 " << w1 << " w2 " << w2 << " b " << b << endl;
        epoch++;
    }
}
int main()
{
    int b, w1, w2, theta, alpha, option;
    cout << "Choose Option " << endl;
    cout << "1. AND" << endl;
    cout << "2. OR " << endl;
    cout << "3. ANDNOT " << endl;
    cout << "4. NAND" << endl;
    cin >> option;
    cout << "give the value of b,w1,w2,theta,alpha" << endl;
    cin >> b >> w1 >> w2 >> theta >> alpha;
    if (option == 1)
    {
        solve(x, y, R1, theta, alpha, b, w1, w2);
    }
    else if (option == 2)
    {
        solve(x, y, R3, theta, alpha, b, w1, w2);
    }
    else if (option == 3)
    {
        solve(x, y, R2, theta, alpha, b, w1, w2);
    }
    else
    {
        solve(x, y, R4, theta, alpha, b, w1, w2);
    }
}